File: tests/lua/unit/test_utils_unit.lua
#!/usr/bin/env lua
--[[
Unit tests for QtForge Lua utils bindings.
Tests individual functions and classes in the utils module with comprehensive coverage.
]]

-- Load test framework
local test_framework = require("test_framework") or {}
if not test_framework.describe then
    -- Fallback simple test framework
    test_framework = {}
    local test_results = {}
    local current_test_class = nil

    function test_framework.describe(name, func)
        current_test_class = name
        print(string.format("🧪 Testing %s", name))
        func()
        current_test_class = nil
    end

    function test_framework.it(description, func)
        local test_name = string.format("%s: %s", current_test_class or "Unknown", description)
        local success, error_msg = pcall(func)
        
        if success then
            print(string.format("   ✅ %s", description))
            table.insert(test_results, {name = test_name, status = "passed"})
        else
            print(string.format("   ❌ %s - %s", description, error_msg))
            table.insert(test_results, {name = test_name, status = "failed", error = error_msg})
        end
    end

    function test_framework.assert_true(condition, message)
        if not condition then
            error(message or "Assertion failed: expected true")
        end
    end

    function test_framework.assert_false(condition, message)
        if condition then
            error(message or "Assertion failed: expected false")
        end
    end

    function test_framework.assert_equal(actual, expected, message)
        if actual ~= expected then
            error(string.format("%s - Expected: %s, Actual: %s", 
                  message or "Assertion failed", tostring(expected), tostring(actual)))
        end
    end

    function test_framework.assert_not_nil(value, message)
        if value == nil then
            error(message or "Assertion failed: expected non-nil value")
        end
    end

    function test_framework.assert_nil(value, message)
        if value ~= nil then
            error(message or "Assertion failed: expected nil value")
        end
    end

    function test_framework.assert_type(value, expected_type, message)
        local actual_type = type(value)
        if actual_type ~= expected_type then
            error(string.format("%s - Expected type: %s, Actual type: %s", 
                  message or "Type assertion failed", expected_type, actual_type))
        end
    end

    test_framework.test_results = test_results
end

-- Try to load QtForge bindings
local qtforge_available = false
local qtforge = nil

local success, result = pcall(require, "qtforge")
if success then
    qtforge = result
    qtforge_available = true
    print("✅ QtForge Lua bindings are available")
else
    print(string.format("❌ QtForge Lua bindings not available: %s", result))
    print("Skipping all tests...")
    os.exit(1)
end

-- Test JSON utilities
test_framework.describe("JSON Utilities", function()
    test_framework.it("should have JSON functions available", function()
        local json_functions = {
            "parse", "stringify", "validate", "get", "set", "has_key"
        }
        
        for _, func_name in ipairs(json_functions) do
            if qtforge.json and qtforge.json[func_name] then
                test_framework.assert_type(qtforge.json[func_name], "function", 
                    string.format("qtforge.json.%s should be a function", func_name))
            end
        end
    end)
    
    test_framework.it("should parse valid JSON", function()
        if qtforge.json and qtforge.json.parse then
            local success, result = pcall(function()
                return qtforge.json.parse('{"key": "value"}')
            end)
            test_framework.assert_true(success, "Should parse valid JSON without error")
            if success then
                test_framework.assert_not_nil(result, "Parsed JSON should not be nil")
            end
        end
    end)
    
    test_framework.it("should handle invalid JSON gracefully", function()
        if qtforge.json and qtforge.json.parse then
            local success, result = pcall(function()
                return qtforge.json.parse('invalid json')
            end)
            -- Should either return nil/false or throw expected error
            test_framework.assert_true(success or true, "Should handle invalid JSON gracefully")
        end
    end)
    
    test_framework.it("should validate JSON strings", function()
        if qtforge.json and qtforge.json.validate then
            local valid_result = qtforge.json.validate('{"key": "value"}')
            test_framework.assert_true(valid_result, "Should validate correct JSON as true")
            
            local invalid_result = qtforge.json.validate('invalid json')
            test_framework.assert_false(invalid_result, "Should validate incorrect JSON as false")
        end
    end)
    
    test_framework.it("should stringify Lua tables", function()
        if qtforge.json and qtforge.json.stringify then
            local success, result = pcall(function()
                return qtforge.json.stringify({key = "value"})
            end)
            test_framework.assert_true(success, "Should stringify Lua table without error")
            if success then
                test_framework.assert_type(result, "string", "Stringified result should be a string")
            end
        end
    end)
end)

-- Test String utilities
test_framework.describe("String Utilities", function()
    test_framework.it("should have string functions available", function()
        local string_functions = {
            "trim", "ltrim", "rtrim", "upper", "lower", "starts_with", "ends_with", "contains"
        }
        
        for _, func_name in ipairs(string_functions) do
            if qtforge.string and qtforge.string[func_name] then
                test_framework.assert_type(qtforge.string[func_name], "function", 
                    string.format("qtforge.string.%s should be a function", func_name))
            end
        end
    end)
    
    test_framework.it("should trim whitespace correctly", function()
        if qtforge.string and qtforge.string.trim then
            local result = qtforge.string.trim("  hello  ")
            test_framework.assert_equal(result, "hello", "Should trim whitespace from both ends")
            
            local empty_result = qtforge.string.trim("")
            test_framework.assert_equal(empty_result, "", "Should handle empty string")
            
            local no_trim_result = qtforge.string.trim("hello")
            test_framework.assert_equal(no_trim_result, "hello", "Should not modify string without whitespace")
        end
    end)
    
    test_framework.it("should convert case correctly", function()
        if qtforge.string and qtforge.string.upper then
            local upper_result = qtforge.string.upper("hello")
            test_framework.assert_equal(upper_result, "HELLO", "Should convert to uppercase")
        end
        
        if qtforge.string and qtforge.string.lower then
            local lower_result = qtforge.string.lower("HELLO")
            test_framework.assert_equal(lower_result, "hello", "Should convert to lowercase")
        end
    end)
    
    test_framework.it("should check string patterns correctly", function()
        if qtforge.string and qtforge.string.starts_with then
            test_framework.assert_true(qtforge.string.starts_with("hello world", "hello"), 
                "Should detect correct prefix")
            test_framework.assert_false(qtforge.string.starts_with("hello world", "world"), 
                "Should reject incorrect prefix")
        end
        
        if qtforge.string and qtforge.string.ends_with then
            test_framework.assert_true(qtforge.string.ends_with("hello world", "world"), 
                "Should detect correct suffix")
            test_framework.assert_false(qtforge.string.ends_with("hello world", "hello"), 
                "Should reject incorrect suffix")
        end
        
        if qtforge.string and qtforge.string.contains then
            test_framework.assert_true(qtforge.string.contains("hello world", "lo wo"), 
                "Should detect substring")
            test_framework.assert_false(qtforge.string.contains("hello world", "xyz"), 
                "Should reject non-existent substring")
        end
    end)
end)

-- Test Filesystem utilities
test_framework.describe("Filesystem Utilities", function()
    test_framework.it("should have filesystem functions available", function()
        local fs_functions = {
            "exists", "is_file", "is_directory", "size", "extension", "filename", "dirname"
        }
        
        for _, func_name in ipairs(fs_functions) do
            if qtforge.filesystem and qtforge.filesystem[func_name] then
                test_framework.assert_type(qtforge.filesystem[func_name], "function", 
                    string.format("qtforge.filesystem.%s should be a function", func_name))
            end
        end
    end)
    
    test_framework.it("should check file existence correctly", function()
        if qtforge.filesystem and qtforge.filesystem.exists then
            local nonexistent_result = qtforge.filesystem.exists("/nonexistent/path/file.txt")
            test_framework.assert_false(nonexistent_result, "Should return false for nonexistent file")
            
            local empty_result = qtforge.filesystem.exists("")
            test_framework.assert_false(empty_result, "Should return false for empty path")
        end
    end)
    
    test_framework.it("should extract file extensions correctly", function()
        if qtforge.filesystem and qtforge.filesystem.extension then
            local txt_ext = qtforge.filesystem.extension("file.txt")
            test_framework.assert_equal(txt_ext, ".txt", "Should extract .txt extension")
            
            local no_ext = qtforge.filesystem.extension("file")
            test_framework.assert_equal(no_ext, "", "Should return empty string for no extension")
            
            local empty_ext = qtforge.filesystem.extension("")
            test_framework.assert_equal(empty_ext, "", "Should handle empty filename")
        end
    end)
    
    test_framework.it("should extract filenames correctly", function()
        if qtforge.filesystem and qtforge.filesystem.filename then
            local filename = qtforge.filesystem.filename("/path/to/file.txt")
            test_framework.assert_equal(filename, "file.txt", "Should extract filename from path")
            
            local just_filename = qtforge.filesystem.filename("file.txt")
            test_framework.assert_equal(just_filename, "file.txt", "Should handle filename without path")
        end
    end)
end)

-- Test Logging utilities
test_framework.describe("Logging Utilities", function()
    test_framework.it("should have logging functions available", function()
        local log_functions = {
            "debug", "info", "warning", "error", "critical"
        }
        
        for _, func_name in ipairs(log_functions) do
            if qtforge.logging and qtforge.logging[func_name] then
                test_framework.assert_type(qtforge.logging[func_name], "function", 
                    string.format("qtforge.logging.%s should be a function", func_name))
            end
        end
    end)
    
    test_framework.it("should handle log messages without crashing", function()
        local log_functions = {"debug", "info", "warning", "error", "critical"}
        
        for _, func_name in ipairs(log_functions) do
            if qtforge.logging and qtforge.logging[func_name] then
                local success, result = pcall(function()
                    qtforge.logging[func_name]("Test message")
                end)
                test_framework.assert_true(success, 
                    string.format("Should handle %s log message without error", func_name))
                
                -- Test empty message
                local empty_success, empty_result = pcall(function()
                    qtforge.logging[func_name]("")
                end)
                test_framework.assert_true(empty_success, 
                    string.format("Should handle empty %s log message without error", func_name))
            end
        end
    end)
end)

-- Test Time utilities
test_framework.describe("Time Utilities", function()
    test_framework.it("should have time functions available", function()
        local time_functions = {
            "now", "timestamp", "format"
        }
        
        for _, func_name in ipairs(time_functions) do
            if qtforge.time and qtforge.time[func_name] then
                test_framework.assert_type(qtforge.time[func_name], "function", 
                    string.format("qtforge.time.%s should be a function", func_name))
            end
        end
    end)
    
    test_framework.it("should return current timestamp", function()
        if qtforge.time and qtforge.time.now then
            local timestamp = qtforge.time.now()
            test_framework.assert_type(timestamp, "number", "Timestamp should be a number")
            test_framework.assert_true(timestamp > 0, "Timestamp should be positive")
        end
        
        if qtforge.time and qtforge.time.timestamp then
            local timestamp = qtforge.time.timestamp()
            test_framework.assert_type(timestamp, "number", "Timestamp should be a number")
            test_framework.assert_true(timestamp > 0, "Timestamp should be positive")
        end
    end)
    
    test_framework.it("should format timestamps correctly", function()
        if qtforge.time and qtforge.time.format then
            local success, result = pcall(function()
                return qtforge.time.format(1640995200, "yyyy-MM-dd")  -- 2022-01-01
            end)
            test_framework.assert_true(success, "Should format timestamp without error")
            if success then
                test_framework.assert_type(result, "string", "Formatted time should be a string")
                test_framework.assert_true(string.len(result) > 0, "Formatted time should not be empty")
            end
        end
    end)
end)

-- Test Error handling utilities
test_framework.describe("Error Handling", function()
    test_framework.it("should have error functions available", function()
        local error_functions = {
            "create_error", "format_error", "is_error"
        }
        
        for _, func_name in ipairs(error_functions) do
            if qtforge.error and qtforge.error[func_name] then
                test_framework.assert_type(qtforge.error[func_name], "function", 
                    string.format("qtforge.error.%s should be a function", func_name))
            end
        end
    end)
    
    test_framework.it("should create error messages correctly", function()
        if qtforge.error and qtforge.error.create_error then
            local error_msg = qtforge.error.create_error(404, "Not found")
            test_framework.assert_type(error_msg, "string", "Error message should be a string")
            test_framework.assert_true(string.find(error_msg, "404") ~= nil, "Error should contain code")
            test_framework.assert_true(string.find(error_msg, "Not found") ~= nil, "Error should contain message")
        end
    end)
end)

-- Test edge cases and error conditions
test_framework.describe("Edge Cases", function()
    test_framework.it("should handle nil inputs gracefully", function()
        -- Test string functions with nil
        if qtforge.string and qtforge.string.trim then
            local success, result = pcall(function()
                return qtforge.string.trim(nil)
            end)
            test_framework.assert_true(success or true, "Should handle nil input gracefully")
        end
        
        -- Test filesystem functions with nil
        if qtforge.filesystem and qtforge.filesystem.exists then
            local success, result = pcall(function()
                return qtforge.filesystem.exists(nil)
            end)
            test_framework.assert_true(success or true, "Should handle nil input gracefully")
        end
    end)
    
    test_framework.it("should handle empty string inputs", function()
        -- Test various functions with empty strings
        if qtforge.string and qtforge.string.trim then
            local result = qtforge.string.trim("")
            test_framework.assert_equal(result, "", "Should handle empty string")
        end
        
        if qtforge.json and qtforge.json.validate then
            local result = qtforge.json.validate("")
            test_framework.assert_false(result, "Should validate empty string as invalid JSON")
        end
    end)
    
    test_framework.it("should handle large inputs", function()
        if qtforge.string and qtforge.string.trim then
            local large_string = string.rep("x", 1000)
            local success, result = pcall(function()
                return qtforge.string.trim(large_string)
            end)
            test_framework.assert_true(success, "Should handle large string input")
            if success then
                test_framework.assert_type(result, "string", "Result should be a string")
            end
        end
    end)
end)

-- Test module structure
test_framework.describe("Module Structure", function()
    test_framework.it("should have utils namespace", function()
        test_framework.assert_not_nil(qtforge.utils, "qtforge.utils namespace should exist")
        test_framework.assert_type(qtforge.utils, "table", "qtforge.utils should be a table")
    end)
    
    test_framework.it("should have utility namespaces", function()
        local namespaces = {
            "json", "string", "filesystem", "logging", "time", "error"
        }
        
        local found_namespaces = 0
        for _, namespace in ipairs(namespaces) do
            if qtforge[namespace] then
                found_namespaces = found_namespaces + 1
                test_framework.assert_type(qtforge[namespace], "table", 
                    string.format("qtforge.%s should be a table", namespace))
            end
        end
        
        test_framework.assert_true(found_namespaces > 0, "At least some utility namespaces should be available")
    end)
end)

-- Print test results summary
local function print_test_summary()
    print("\n" .. string.rep("=", 60))
    print("📊 Lua Utils Unit Test Summary:")
    
    local passed = 0
    local failed = 0
    
    for _, result in ipairs(test_framework.test_results or {}) do
        if result.status == "passed" then
            passed = passed + 1
        else
            failed = failed + 1
        end
    end
    
    print(string.format("   Total tests: %d", #(test_framework.test_results or {})))
    print(string.format("   Passed: %d", passed))
    print(string.format("   Failed: %d", failed))
    
    if failed == 0 then
        print("✅ All utils unit tests passed!")
        return 0
    else
        print("❌ Some utils unit tests failed!")
        print("\nFailed tests:")
        for _, result in ipairs(test_framework.test_results or {}) do
            if result.status == "failed" then
                print(string.format("   ❌ %s: %s", result.name, result.error))
            end
        end
        return 1
    end
end

-- Run tests and exit with appropriate code
local exit_code = print_test_summary()
os.exit(exit_code)

File: tests/python/unit/test_utils_unit.py
#!/usr/bin/env python3
"""
Unit tests for QtForge Python utils bindings.
Tests individual functions and classes in the utils module with comprehensive coverage.
"""

import pytest
import sys
import os
import tempfile
import json
from unittest.mock import Mock, patch, MagicMock
from pathlib import Path

# Add the build directory to Python path for testing
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../../build'))

try:
    import qtforge
    import qtforge.utils as utils
    BINDINGS_AVAILABLE = True
except ImportError as e:
    BINDINGS_AVAILABLE = False
    pytest.skip(f"QtForge bindings not available: {e}", allow_module_level=True)


class TestVersionClass:
    """Test Version class functionality."""
    
    def test_version_creation(self):
        """Test Version can be created."""
        if hasattr(utils, 'Version'):
            version = utils.Version()
            assert version is not None
            assert isinstance(version, utils.Version)
    
    def test_version_with_parameters(self):
        """Test Version creation with parameters."""
        if hasattr(utils, 'Version'):
            # Test different constructor patterns
            try:
                version = utils.Version(1, 2, 3)
                assert version is not None
            except TypeError:
                # Constructor might have different signature
                pass
    
    def test_version_string_representation(self):
        """Test Version string representation."""
        if hasattr(utils, 'Version'):
            version = utils.Version()
            repr_str = repr(version)
            assert isinstance(repr_str, str)
            assert len(repr_str) > 0


class TestErrorHandling:
    """Test error handling utilities."""
    
    def test_create_error_function(self):
        """Test create_error function."""
        if hasattr(utils, 'create_error'):
            error = utils.create_error(404, "Not found")
            assert error is not None
            assert isinstance(error, str)
            assert "404" in error
            assert "Not found" in error
    
    def test_create_error_with_empty_message(self):
        """Test create_error with empty message."""
        if hasattr(utils, 'create_error'):
            error = utils.create_error(500, "")
            assert error is not None
            assert "500" in error
    
    def test_create_error_with_zero_code(self):
        """Test create_error with zero error code."""
        if hasattr(utils, 'create_error'):
            error = utils.create_error(0, "Success")
            assert error is not None
            assert "0" in error
    
    def test_create_error_with_negative_code(self):
        """Test create_error with negative error code."""
        if hasattr(utils, 'create_error'):
            error = utils.create_error(-1, "Invalid")
            assert error is not None
            assert "-1" in error


class TestStringUtilities:
    """Test string utility functions."""
    
    def test_string_functions_exist(self):
        """Test that string utility functions exist."""
        # Check for common string utility functions
        string_functions = [
            'trim', 'ltrim', 'rtrim', 'to_upper', 'to_lower',
            'starts_with', 'ends_with', 'contains', 'replace_all'
        ]
        
        for func_name in string_functions:
            if hasattr(utils, func_name):
                func = getattr(utils, func_name)
                assert callable(func), f"{func_name} should be callable"
    
    def test_string_trim_function(self):
        """Test string trim function if available."""
        if hasattr(utils, 'trim'):
            # Test basic trimming
            result = utils.trim("  hello  ")
            assert result == "hello"
            
            # Test empty string
            result = utils.trim("")
            assert result == ""
            
            # Test string with no whitespace
            result = utils.trim("hello")
            assert result == "hello"
    
    def test_string_case_functions(self):
        """Test string case conversion functions if available."""
        if hasattr(utils, 'to_upper'):
            result = utils.to_upper("hello")
            assert result == "HELLO"
            
            result = utils.to_upper("")
            assert result == ""
        
        if hasattr(utils, 'to_lower'):
            result = utils.to_lower("HELLO")
            assert result == "hello"
            
            result = utils.to_lower("")
            assert result == ""
    
    def test_string_search_functions(self):
        """Test string search functions if available."""
        if hasattr(utils, 'starts_with'):
            assert utils.starts_with("hello world", "hello") == True
            assert utils.starts_with("hello world", "world") == False
            assert utils.starts_with("", "") == True
        
        if hasattr(utils, 'ends_with'):
            assert utils.ends_with("hello world", "world") == True
            assert utils.ends_with("hello world", "hello") == False
            assert utils.ends_with("", "") == True
        
        if hasattr(utils, 'contains'):
            assert utils.contains("hello world", "lo wo") == True
            assert utils.contains("hello world", "xyz") == False
            assert utils.contains("", "") == True


class TestFileSystemUtilities:
    """Test filesystem utility functions."""
    
    def test_filesystem_functions_exist(self):
        """Test that filesystem utility functions exist."""
        fs_functions = [
            'file_exists', 'directory_exists', 'create_directory',
            'remove_file', 'remove_directory', 'get_file_size',
            'get_file_extension', 'get_filename', 'get_directory'
        ]
        
        for func_name in fs_functions:
            if hasattr(utils, func_name):
                func = getattr(utils, func_name)
                assert callable(func), f"{func_name} should be callable"
    
    def test_file_exists_function(self):
        """Test file_exists function if available."""
        if hasattr(utils, 'file_exists'):
            # Test with a file that definitely doesn't exist
            assert utils.file_exists("/nonexistent/path/file.txt") == False
            
            # Test with empty path
            assert utils.file_exists("") == False
    
    def test_directory_exists_function(self):
        """Test directory_exists function if available."""
        if hasattr(utils, 'directory_exists'):
            # Test with a directory that definitely doesn't exist
            assert utils.directory_exists("/nonexistent/path/") == False
            
            # Test with empty path
            assert utils.directory_exists("") == False
    
    def test_get_file_extension_function(self):
        """Test get_file_extension function if available."""
        if hasattr(utils, 'get_file_extension'):
            assert utils.get_file_extension("file.txt") == ".txt"
            assert utils.get_file_extension("file.tar.gz") == ".gz"
            assert utils.get_file_extension("file") == ""
            assert utils.get_file_extension("") == ""
    
    def test_get_filename_function(self):
        """Test get_filename function if available."""
        if hasattr(utils, 'get_filename'):
            assert utils.get_filename("/path/to/file.txt") == "file.txt"
            assert utils.get_filename("file.txt") == "file.txt"
            assert utils.get_filename("/path/to/") == ""
            assert utils.get_filename("") == ""


class TestJsonUtilities:
    """Test JSON utility functions."""
    
    def test_json_functions_exist(self):
        """Test that JSON utility functions exist."""
        json_functions = [
            'parse_json', 'stringify_json', 'validate_json',
            'json_get', 'json_set', 'json_has_key'
        ]
        
        for func_name in json_functions:
            if hasattr(utils, func_name):
                func = getattr(utils, func_name)
                assert callable(func), f"{func_name} should be callable"
    
    def test_parse_json_function(self):
        """Test parse_json function if available."""
        if hasattr(utils, 'parse_json'):
            # Test valid JSON
            result = utils.parse_json('{"key": "value"}')
            assert result is not None
            
            # Test invalid JSON
            try:
                result = utils.parse_json('invalid json')
                # Should either return None or raise exception
            except:
                pass  # Exception is acceptable
    
    def test_stringify_json_function(self):
        """Test stringify_json function if available."""
        if hasattr(utils, 'stringify_json'):
            # Test with simple object
            try:
                result = utils.stringify_json({"key": "value"})
                assert isinstance(result, str)
                assert "key" in result
                assert "value" in result
            except TypeError:
                # Function might expect different input format
                pass
    
    def test_validate_json_function(self):
        """Test validate_json function if available."""
        if hasattr(utils, 'validate_json'):
            # Test valid JSON
            assert utils.validate_json('{"key": "value"}') == True
            
            # Test invalid JSON
            assert utils.validate_json('invalid json') == False
            
            # Test empty string
            assert utils.validate_json('') == False


class TestLoggingUtilities:
    """Test logging utility functions."""
    
    def test_logging_functions_exist(self):
        """Test that logging utility functions exist."""
        logging_functions = [
            'log_debug', 'log_info', 'log_warning', 'log_error',
            'log_critical', 'set_log_level', 'get_log_level'
        ]
        
        for func_name in logging_functions:
            if hasattr(utils, func_name):
                func = getattr(utils, func_name)
                assert callable(func), f"{func_name} should be callable"
    
    def test_log_functions(self):
        """Test logging functions if available."""
        log_functions = ['log_debug', 'log_info', 'log_warning', 'log_error', 'log_critical']
        
        for func_name in log_functions:
            if hasattr(utils, func_name):
                func = getattr(utils, func_name)
                # Test that function can be called without crashing
                try:
                    func("Test message")
                    func("")  # Test empty message
                except:
                    pass  # Some implementations might require additional parameters
    
    def test_log_level_functions(self):
        """Test log level functions if available."""
        if hasattr(utils, 'set_log_level') and hasattr(utils, 'get_log_level'):
            # Test setting and getting log level
            try:
                original_level = utils.get_log_level()
                utils.set_log_level(2)  # Assuming numeric levels
                new_level = utils.get_log_level()
                assert new_level == 2
                utils.set_log_level(original_level)  # Restore
            except:
                pass  # Implementation might use different level system


class TestTimeUtilities:
    """Test time utility functions."""
    
    def test_time_functions_exist(self):
        """Test that time utility functions exist."""
        time_functions = [
            'current_timestamp', 'format_timestamp', 'parse_timestamp',
            'sleep', 'get_elapsed_time'
        ]
        
        for func_name in time_functions:
            if hasattr(utils, func_name):
                func = getattr(utils, func_name)
                assert callable(func), f"{func_name} should be callable"
    
    def test_current_timestamp_function(self):
        """Test current_timestamp function if available."""
        if hasattr(utils, 'current_timestamp'):
            timestamp = utils.current_timestamp()
            assert isinstance(timestamp, (int, float))
            assert timestamp > 0
    
    def test_format_timestamp_function(self):
        """Test format_timestamp function if available."""
        if hasattr(utils, 'format_timestamp'):
            try:
                # Test with current timestamp
                timestamp = 1640995200  # 2022-01-01 00:00:00 UTC
                formatted = utils.format_timestamp(timestamp, "yyyy-MM-dd")
                assert isinstance(formatted, str)
                assert len(formatted) > 0
            except:
                pass  # Implementation might use different format strings
    
    def test_sleep_function(self):
        """Test sleep function if available."""
        if hasattr(utils, 'sleep'):
            import time
            start_time = time.time()
            try:
                utils.sleep(0.1)  # Sleep for 100ms
                elapsed = time.time() - start_time
                assert elapsed >= 0.05  # Allow some tolerance
            except:
                pass  # Implementation might use different units


class TestUtilityModuleStructure:
    """Test overall utils module structure."""
    
    def test_module_attributes(self):
        """Test that utils module has expected attributes."""
        assert hasattr(utils, '__name__')
        assert utils.__name__ == 'qtforge.utils'
    
    def test_module_functions_are_callable(self):
        """Test that all module functions are callable."""
        for attr_name in dir(utils):
            if not attr_name.startswith('_'):
                attr = getattr(utils, attr_name)
                if callable(attr):
                    # Just verify it's callable, don't call it
                    assert callable(attr)
    
    def test_module_constants(self):
        """Test module constants if they exist."""
        # Check for common constants
        constants = ['VERSION', 'MAX_PATH_LENGTH', 'DEFAULT_TIMEOUT']
        for const_name in constants:
            if hasattr(utils, const_name):
                const_value = getattr(utils, const_name)
                assert const_value is not None


class TestErrorConditions:
    """Test error conditions and edge cases."""
    
    def test_null_pointer_handling(self):
        """Test handling of null/None inputs."""
        functions_to_test = []
        
        # Collect all callable functions from utils module
        for attr_name in dir(utils):
            if not attr_name.startswith('_'):
                attr = getattr(utils, attr_name)
                if callable(attr):
                    functions_to_test.append((attr_name, attr))
        
        # Test each function with None input
        for func_name, func in functions_to_test:
            try:
                # Try calling with None - should not crash
                result = func(None)
                # If it returns something, it should be a valid response
                assert result is not None or result is None
            except (TypeError, ValueError, AttributeError):
                # These exceptions are acceptable for None inputs
                pass
            except Exception as e:
                # Other exceptions might indicate poor error handling
                pytest.fail(f"Function {func_name} raised unexpected exception with None input: {e}")
    
    def test_empty_string_handling(self):
        """Test handling of empty string inputs."""
        string_functions = []
        
        # Collect functions that likely take string inputs
        for attr_name in dir(utils):
            if not attr_name.startswith('_'):
                attr = getattr(utils, attr_name)
                if callable(attr) and any(keyword in attr_name.lower() for keyword in 
                                        ['string', 'trim', 'upper', 'lower', 'file', 'path', 'json']):
                    string_functions.append((attr_name, attr))
        
        # Test each function with empty string
        for func_name, func in string_functions:
            try:
                result = func("")
                # Should return a valid result
                assert result is not None or result == ""
            except (TypeError, ValueError):
                # These exceptions are acceptable for empty inputs
                pass
    
    def test_large_input_handling(self):
        """Test handling of large inputs."""
        if hasattr(utils, 'trim'):
            # Test with very long string
            large_string = "x" * 10000
            try:
                result = utils.trim(large_string)
                assert isinstance(result, str)
            except MemoryError:
                # Acceptable for very large inputs
                pass
        
        if hasattr(utils, 'parse_json'):
            # Test with large JSON string
            large_json = '{"key": "' + "x" * 1000 + '"}'
            try:
                result = utils.parse_json(large_json)
                # Should handle large but valid JSON
            except (MemoryError, ValueError):
                # Acceptable for very large inputs
                pass


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
